komorebi-bar/src/group.rs

```diff
@@ -0,0 +1,161 @@
+                })
+                .show(ui, add_contents),
+            Self::Side(_config) => InnerResponse {
+                inner: add_contents(ui),
+                response: ui.response().clone(),
+            },
+            Self::None => InnerResponse {
+                inner: add_contents(ui),
+                response: ui.response().clone(),
+            },
+        }
+    }
+
+    pub fn apply_on_side<R>(
+        &mut self,
+        ui: &mut Ui,
+        add_contents: impl FnOnce(&mut Ui) -> R,
+    ) -> InnerResponse<R> {
+        match self {
+            Self::Widget(_config) => InnerResponse {
+                inner: add_contents(ui),
+                response: ui.response().clone(),
+            },
+            Self::Side(config) => Frame::none()
+                .fill(match config.fill {
+                    Some(color) => color.into(),
+                    None => Color32::TRANSPARENT,
+                })
+                .outer_margin(match config.outer_margin {
+                    Some(margin) => Self::rect_to_margin(margin),
+                    None => Margin::symmetric(0.0, 0.0),
+                })
+                .inner_margin(match config.inner_margin {
+                    Some(margin) => Self::rect_to_margin(margin),
+                    None => Margin::symmetric(5.0, 2.0),
+                })
+                .rounding(match config.rounding {
+                    Some(rounding) => rounding.into(),
+                    None => Rounding::same(5.0),
+                })
+                .stroke(match config.stroke {
+                    Some(line) => line.into(),
+                    None => ui.style().visuals.widgets.noninteractive.bg_stroke,
+                })
+                .show(ui, add_contents),
+            Self::None => InnerResponse {
+                inner: add_contents(ui),
+                response: ui.response().clone(),
+            },
+        }
+    }
+
+    fn rect_to_margin(rect: Rect) -> Margin {
+        Margin {
+            left: rect.left as f32,
+            right: rect.right as f32,
+            top: rect.top as f32,
+            bottom: rect.bottom as f32,
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, Serialize, Deserialize, JsonSchema)]
+pub struct GroupingConfig {
+    pub fill: Option<AlphaColour>,
+    pub rounding: Option<RoundingConfig>,
+    pub outer_margin: Option<Rect>,
+    pub inner_margin: Option<Rect>,
+    pub stroke: Option<Line>,
+}
+
+#[derive(Copy, Clone, Debug, Serialize, Deserialize, JsonSchema)]
+pub struct Line {
+    pub width: f32,
+    pub color: Option<Colour>,
+}
+
+impl From<Line> for Stroke {
+    fn from(value: Line) -> Self {
+        Self {
+            width: value.width,
+            color: match value.color {
+                Some(color) => color.into(),
+                None => Color32::from_rgb(0, 0, 0),
+            },
+        }
+    }
+}
+
+#[derive(Copy, Clone, Debug, Serialize, Deserialize, JsonSchema)]
+pub enum RoundingConfig {
```

Comment:
Let's make this #[serde(untagged)] so that it's possible to specify:

```json
rounding: 5.0,
```

or

```json
rounding: [5.0, 5.0, 5.0, 5.0]
```

Right now `grouping.Widget.Same` and `grouping.Widget.Individual` feels very clunky and unintuitive

Reply:
It's bedtime for me (2am). I will look into these later. Thank you for your feedback on this!

Reply:
All in your own time, there is no rush!